### PR TITLE
Workflow for tracking PRs assignment

### PR DESCRIPTION
**Pull request by apiraino**
This is the third try :sweat_smile: to land part of the work on #1753 (pull requests tracking and balanced assignment for the Rust project contributors).

This patch implements:
- Added a new DB table with the fields to track how many PRs are assigned to a contributor
- Initial DB table population with a one-off job, manually run.

r? @jackh726

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -261,6 +262,12 @@ async fn run_server(addr: SocketAddr) -> anyhow::Result<()> {
         octocrab: oc,
     });
 
+    // Run all jobs that don't have a schedule (one-off jobs)
+    if !is_scheduled_jobs_disabled() {
+        spawn_job_oneoffs(ctx.clone());
+    }
```
Comment (apiraino) [inline src/main.rs:268]: as a stopgap solution I've put the job trigger here

File: src/main.rs [hunk 2]
```diff
@@ -310,6 +317,38 @@ async fn run_server(addr: SocketAddr) -> anyhow::Result<()> {
     Ok(())
 }
 
+/// Spawns a background tokio task which runs all jobs having no schedule
+/// i.e. manually started
+fn spawn_job_oneoffs(ctx: Arc<Context>) {
+    // TODO: Ideally JobSchedule.schedule should become an `Option<Schedule>`
+    // and here we run all those with schedule=None
```
Comment (apiraino) [inline src/main.rs:324]: @jackh726 annoted here your suggestion on how to run non-scheduled job (for a separate patch)

**Comment by apiraino** [comment]
The job starts but GH returns an error, so I'll check what's happening :slightly_smiling_face:

File: github-graphql/PullRequestsOpen.gql [hunk 3]
```diff
@@ -0,0 +1,26 @@
+query PullRequestsOpen ($repo_owner: String!, $repo_name: String!, $after: String) {
+  repository(owner: $repo_owner, name: $repo_name) {
+    pullRequests(first: 100, after: $after, states:OPEN, labels: ["S-waiting-on-review","T-compiler"]) {
```
Comment (jackh726) [inline github-graphql/PullRequestsOpen.gql:3]: This should not be limited to these labels.

File: github-graphql/PullRequestsOpen.gql [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        number
+        updatedAt
+        createdAt
+        assignees(first: 10) {
+          nodes {
+            login
+            id
+          }
+        }
+        labels(first:5, orderBy:{field:NAME, direction:DESC}) {
```
Comment (jackh726) [inline github-graphql/PullRequestsOpen.gql:18]: Looking at this, I assume you later use this to filter labels, but I don't think we should do that.

File: github-graphql/src/lib.rs [hunk 5]
```diff
@@ -385,3 +386,44 @@ pub mod project_items {
...
+    #[cynic(graphql_type = "Query", variables = "PullRequestsOpenVariables")]
+    pub struct PullRequestsOpen {
+        #[arguments(owner: $repo_owner, name: $repo_name)]
+        pub repository: Option<Repository>,
+    }
+
+    #[derive(cynic::QueryFragment, Debug)]
+    #[cynic(variables = "PullRequestsOpenVariables")]
+    pub struct Repository {
+        #[arguments(first: 100, after: $after, states: "OPEN", labels: ["S-waiting-on-review","T-compiler"])]
```
Comment (jackh726) [inline github-graphql/src/lib.rs:415]: Are these needed? Seems redundant with the gql file.

Comment (jackh726) [reply github-graphql/src/lib.rs:415]: Still, like I commented above, I don't think we should limit this.

File: src/github.rs [hunk 6]
```diff
@@ -2592,6 +2592,81 @@ async fn project_items_by_status(
     Ok(all_items)
 }
 
+/// Retrieve all pull requests in status OPEN, not DRAFT, with label: S-waiting-on-review
```
Comment (jackh726) [inline src/github.rs:2595]: Same as above.

File: src/handlers/pull_requests_assignment_update.rs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    async fn run(&self, ctx: &super::Context, _metadata: &serde_json::Value) -> anyhow::Result<()> {
+        let db = ctx.db.get().await;
+        let gh = &ctx.github;
+
+        tracing::trace!("starting pull_request_assignment_update");
+
+        // delete everything before populating
+        init_table(&db).await?;
```
Comment (jackh726) [inline src/handlers/pull_requests_assignment_update.rs:22]: This should be moved to the end: if we error out, we don't want to lose everything.

File: src/handlers/pull_requests_assignment_update.rs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+        let rust_repo = gh.repository("rust-lang/rust").await?;
+        let prs = retrieve_pull_requests(&rust_repo, &gh).await?;
+
+        // populate the table
+        for (assignees, pr_num) in prs {
+            for assignee in assignees {
+                let assignee_id = assignee.id.expect("checked");
```
Comment (jackh726) [inline src/handlers/pull_requests_assignment_update.rs:30]: Preexisting (but maybe you can make a separate PR for this?): id on User probably shouldn't be optional.

File: src/handlers/pull_requests_assignment_update.rs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Truncate the review prefs table
+async fn init_table(db: &DbClient) -> anyhow::Result<u64> {
+    let res = db.execute("TRUNCATE review_prefs;", &[]).await?;
```
Comment (jackh726) [inline src/handlers/pull_requests_assignment_update.rs:42]: I don't think this is right. We don't want to delete everything, just the assigned reviews.

File: src/handlers/pull_requests_assignment_update.rs [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        .await
+        .context("Insert user DB error")?;
+    Ok(rec)
+}
+
+/// Create a team member work queue
+async fn create_team_member_workqueue(
+    db: &DbClient,
+    user_id: i64,
+    pr: i32,
```
Comment (jackh726) [inline src/handlers/pull_requests_assignment_update.rs:62]: I feel like we should aggregate by user *first*, rather than doing a db request for each PR.

File: src/main.rs [hunk 11]
```diff
@@ -310,6 +317,38 @@ async fn run_server(addr: SocketAddr) -> anyhow::Result<()> {
...
+    let jobs: Vec<Box<dyn Job + Send + Sync>> = vec![Box::new(PullRequestAssignmentUpdate)];
+
+    for job in jobs {
+        let ctx = ctx.clone();
+        task::spawn(async move {
+            // Allow some spacing between starting jobs
+            let mut interval =
```
Comment (jackh726) [inline src/main.rs:331]: Shouldn't this be moved outside the loop? So each job is delayed between each other?

File: src/main.rs (hunk 2 above)
Comment (jackh726) [reply src/main.rs:324]: Hmm, not quite what I was thinking, but this is good!

File: src/main.rs [hunk 12]
```diff
@@ -261,6 +262,12 @@ async fn run_server(addr: SocketAddr) -> anyhow::Result<()> {
         octocrab: oc,
     });
 
+    // Run all jobs that don't have a schedule (one-off jobs)
+    if !is_scheduled_jobs_disabled() {
```
Comment (jackh726) [inline src/main.rs:266]: Hmm, what is this function

Comment (jackh726) [reply src/main.rs:266]: Oh, its preexisting. Okay.

File: github-graphql/PullRequestsOpen.gql (hunk 3 above)
Comment (apiraino) [reply github-graphql/PullRequestsOpen.gql:3]: hmm you're right, we should pull **all** open PRs here. Don't know what I was thinking...

File: github-graphql/PullRequestsOpen.gql (hunk 4 above)
Comment (apiraino) [reply github-graphql/PullRequestsOpen.gql:18]: right, currently I don't use them but I thought it could make sense to pull them anyway in the query (for future use?). Would it be ok to pull them but **not** use them for filtering?

File: src/handlers/pull_requests_assignment_update.rs (hunk 8 above)
Comment (apiraino) [reply src/handlers/pull_requests_assignment_update.rs:30]: yeah this is something that confuses me, don't know exactly why our https://github.com/rust-lang/triagebot/blob/c52016ab56eddb5ce2ac51a7bde6a47b8653719a/src/github.rs#L19-L23 has the Id optional :shrug: 

I agree to remove the `Option` is a separate PR.

File: src/handlers/pull_requests_assignment_update.rs (hunk 9 above)
Comment (apiraino) [reply src/handlers/pull_requests_assignment_update.rs:42]: do you suggest to simply delete all assignments? What if we have GH users that left team? I am not sure the `users` table of the triagebot is ever purged (didn't check yet).

My reasoning is: when invoking a job to initialize the table I thought it was easier to start from a clean slate. This `init_table` should only be called when the intent is to initialize the table.

No strong opinion, though, about just deleting the assignments. It will just require adding a bit of logic to `create_team_member_workqueue` to turn it into an `upsert`

File: github-graphql/PullRequestsOpen.gql (hunk 4 above)
Comment (jackh726) [reply github-graphql/PullRequestsOpen.gql:18]: That's fine

File: src/handlers/pull_requests_assignment_update.rs (hunk 7 above)
Comment (jackh726) [reply src/handlers/pull_requests_assignment_update.rs:22]: Well, I think this should at least be after the web requests. If those fail for whatever reason, we shouldn't just lose the data (even though it's easier to get back).

File: src/handlers/pull_requests_assignment_update.rs (hunk 9 above)
Comment (jackh726) [reply src/handlers/pull_requests_assignment_update.rs:42]: The table is also supposed to store preferences, right? If we delete table, preferences are lost.

I guess we could periodically cleanup users, but don't think it's that big of an issue at the moment.

File: github-graphql/README.md [hunk 13]
```diff
@@ -0,0 +1,30 @@
+# How to use GraphQL with Rust
```
Comment (apiraino) [inline github-graphql/README.md:1]: I will note down is a separate issue to make a little refactor of the GraphQL code. It has grown a bit organically but now I think we can establish some guidelines on how to do that

File: src/github.rs [hunk 14]
```diff
@@ -2592,6 +2592,93 @@ async fn project_items_by_status(
...
+            let labels = pr
+                .labels
+                .map(|l| l.nodes)
+                .unwrap_or_default()
+                .into_iter()
+                .map(|node| node.name)
+                .collect::<Vec<_>>();
+            if labels.iter().any(|label| label == "rollup") {
+                return None;
+            }
```
Comment (apiraino) [inline src/github.rs:2655]: maybe I've found a good use for the labels we pull in this query: exclude the rollups because they don't follow the same lifecycle of other PRs

File: src/handlers/pull_requests_assignment_update.rs [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+/// Truncate the review prefs table
+async fn init_table(db: &DbClient) -> anyhow::Result<u64> {
+    let res = db
+        .execute("UPDATE review_prefs SET assigned_prs='{}';", &[])
+        .await?;
+    Ok(res)
+}
+
+/// Add a new user (if not existing)
+async fn ensure_team_member(db: &DbClient, user_id: i64, username: &str) -> anyhow::Result<u64> {
```
Comment (jackh726) [inline src/handlers/pull_requests_assignment_update.rs:49]: nit: we should probably de-duplicate this with the existing `notifications::record_username`

**Review by jackh726 — COMMENTED** [review]
Okay, seems very close. There are still a couple open comments from me and a couple new ones.

I don't think anything here is *blocking*. If you can get to it now, good. Otherwise, a TODO will be okay.

**Comment by apiraino** [comment]
> Okay, seems very close. There are still a couple open comments from me and a couple new ones.
> 
> I don't think anything here is _blocking_. If you can get to it now, good. Otherwise, a TODO will be okay.

just e bit more time, I'm working on resolving the open comments because they do make sense :)

File: src/github.rs (hunk 14 above)
Comment (apiraino) [reply src/github.rs:2654]: (good point)

File: src/handlers/pull_requests_assignment_update.rs [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+        init_table(&db).await?;
+
+        // aggregate by user first
+        let aggregated = prs.into_iter().fold(HashMap::new(), |mut acc, (user, pr)| {
+            let (_, prs) = acc
+                .entry(user.id.unwrap())
+                .or_insert_with(|| (user, Vec::new()));
+            prs.push(pr);
+            acc
+        });
```
Comment (apiraino) [inline src/handlers/pull_requests_assignment_update.rs:37]: now pull requests are aggregated by user before executing queries (this snippet was a bit tricky to get right, I was helped!)

File: src/handlers/pull_requests_assignment_update.rs [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+        // populate the table
+        for (_user_id, (assignee, prs)) in &aggregated {
+            let assignee_id = assignee.id.expect("checked");
+            let _ = record_username(&db, assignee_id, &assignee.login).await;
```
Comment (apiraino) [inline src/handlers/pull_requests_assignment_update.rs:42]: I am now using `record_username` and removed the deuplicate function. I've slightly changed the signature of `record_username`.

File: src/handlers/pull_requests_assignment_update.rs (hunk 16 above)
Comment (jackh726) [reply src/handlers/pull_requests_assignment_update.rs:37]: This looks right, but in the future I find it much easier to just use an imperative style (for loop) with a mutable map.

File: src/main.rs (hunk 11 above)
Comment (apiraino) [reply src/main.rs]: ok, now the job runner should space a bit each job execution. Added a comment to clarify that there is no guarantee they start in sequence and that the delay is precise. But I think it should do.

**Comment by apiraino** [comment]
I think all the comments are now resolved :+1: 

@rustbot ready

File: src/main.rs (hunk 12 above)
Comment (jackh726) [reply src/main.rs:268]: I'm not going to block this PR on it, but this should be able to be in the same if block as below. Please fix next PR

File: src/main.rs [hunk 18]
```diff
@@ -310,6 +319,35 @@ async fn run_server(addr: SocketAddr) -> anyhow::Result<()> {
...
+/// i.e. manually executed at the end of the triagebot startup
+// - jobs are not guaranteed to start in sequence (care is to be taken to ensure thet are completely independent one from the other)
+// - the delay between jobs start is not guaranteed to be precise
+async fn spawn_job_oneoffs(ctx: Arc<Context>) {
+    let jobs: Vec<Box<dyn Job + Send + Sync>> = vec![Box::new(PullRequestAssignmentUpdate)];
+
+    for (idx, job) in jobs.into_iter().enumerate() {
+        let ctx = ctx.clone();
+        task::spawn(async move {
+            // Allow some spacing between starting jobs
```
Comment (jackh726) [inline src/main.rs:332]: Also not going to block on this, but seems like we could also just spawn one task that runs through all the jobs. Or just not worry about the delay (we can always add it later if it's a problem).

**Review by jackh726 — APPROVED** [review]
A couple comments that I'd like for you to followup on in another PR, but let's land this.
